.github/workflows/ci.yml

```
@@ -74,3 +78,12 @@ jobs:
           -Psigning.keyId="${{ secrets.SIGNING_KEY_ID }}"
           -Psigning.password="${{ secrets.SIGNING_PASSWORD }}"
           -Psigning.secretKeyRingFile="$HOME/secring.gpg"
+
+      - name: Publish NPM
+        if: github.event_name == 'release' && startsWith(github.ref, 'refs/tags/')
+        run: >-
+          ./gradlew
+          $GRADLE_ARGS
+          publishJsNpmPublicationToGithub
+          -Pnpm.publish.authToken.github="${{ secrets.MOBILE_GITHUB_PACKAGES_PASSWORD }}"
+          -Pnpm.publish.version="${{ steps.get_version.outputs.VERSION }}"
```

Rather than have a separate `step` ("Get the version" above), can we just use the variable directly here?

```suggestion
          -Pnpm.publish.version="${GITHUB_REF/refs\/tags\//}"
```

---

Possibly worth combining the two publish steps into one?

```
      - name: Publish
        if: startsWith(github.ref, 'refs/tags/')
        run: >-
          ./gradlew
          $GRADLE_ARGS
          uploadArchives
          publishJsNpmPublicationToGithub
          -PVERSION_NAME=${GITHUB_REF/refs\/tags\//}
          -Pversion="${GITHUB_REF/refs\/tags\//}"
          -Psigning.keyId="${{ secrets.SIGNING_KEY_ID }}"
          -Psigning.password="${{ secrets.SIGNING_PASSWORD }}"
          -Psigning.secretKeyRingFile="$HOME/secring.gpg"
```

I haven't checked, but the `com.vanniktech.maven.publish` plugin may work with `publish` task, which I believe would allow us to replace `uploadArchives` with `publish` and have it work for both publication to both Maven and npm.

---

Alternatively, it should be simple enough to have `publish` task depend on both the Maven (`installArchives`) and npm (`publishJsNpmPublicationToGithub`).

- https://github.com/gradle/kotlin-dsl-samples/blob/master/samples/task-dependencies/build.gradle.kts
- https://docs.gradle.org/current/userguide/more_about_tasks.html

---

For sure we can use the pattern extraction directly without the extra step, have made that change. Will also use this to override the `project.version` variable as suggested.

As for combining the publishing tasks I'd recommend we keep it separate for the moment until we publish a live tag and verify that everything is operating standalone. Once that is confirmed we can go about optimizing the publishing tasks as a combined sequence in a separate PR.

---

FYI, publishToNpm tasks generated by the plugin already are added as dependencies to publish task :)

---

Quick follow up:

`npm-publish` conveniently makes sure npm tasks are configured as dependencies of `publish`, unfortunately if we want to keep our Maven and npm publications steps separate we'll have to explicitly call out the npm task.

In the future, we might try combining the publication (and use the official maven publish Gradle plugin) at which point we can use the `publish` task.

---

Demonstrating `publish` depending on our Maven publication tasks:

<details>
<summary>Gradle run</summary>

```
./gradlew -Pnpm.publish.authToken.github=foo publish --dry-run

:kotlinNodeJsSetup SKIPPED
:koap:jsPackageJson SKIPPED
:rootPackageJson SKIPPED
:kotlinNpmInstall SKIPPED
:koap:jsGenerateExternalsIntegrated SKIPPED
:koap:compileKotlinJs SKIPPED
:koap:jsProcessResources SKIPPED
:koap:assembleJsNpmPublication SKIPPED
:koap:publishJsNpmPublicationToGithub SKIPPED
:koap:jsMainClasses SKIPPED
:koap:jsPublicPackageJson SKIPPED
:koap:jsJar SKIPPED
:koap:generateMetadataFileForJsPublication SKIPPED
:koap:generatePomFileForJsPublication SKIPPED
:koap:compileJava SKIPPED
:koap:processResources SKIPPED
:koap:classes SKIPPED
:koap:javadoc SKIPPED
:koap:javadocsJar SKIPPED
:koap:jsSourcesJar SKIPPED
:koap:signJsPublication SKIPPED
:koap:publishJsPublicationToLocalRepository SKIPPED
:koap:publishJsPublicationToMavenRepository SKIPPED
:koap:compileKotlinJvm SKIPPED
:koap:jvmProcessResources SKIPPED
:koap:jvmMainClasses SKIPPED
:koap:jvmJar SKIPPED
:koap:generateMetadataFileForJvmPublication SKIPPED
:koap:generatePomFileForJvmPublication SKIPPED
:koap:jvmSourcesJar SKIPPED
:koap:signJvmPublication SKIPPED
:koap:publishJvmPublicationToLocalRepository SKIPPED
:koap:publishJvmPublicationToMavenRepository SKIPPED
:koap:emptySourcesJar SKIPPED
:koap:generateMetadataFileForKotlinMultiplatformPublication SKIPPED
:koap:generatePomFileForKotlinMultiplatformPublication SKIPPED
:koap:signKotlinMultiplatformPublication SKIPPED
:koap:publishKotlinMultiplatformPublicationToLocalRepository SKIPPED
:koap:publishKotlinMultiplatformPublicationToMavenRepository SKIPPED
:koap:compileKotlinMetadata SKIPPED
:koap:metadataMainClasses SKIPPED
:koap:metadataJar SKIPPED
:koap:generateMetadataFileForMetadataPublication SKIPPED
:koap:generatePomFileForMetadataPublication SKIPPED
:koap:metadataSourcesJar SKIPPED
:koap:signMetadataPublication SKIPPED
:koap:publishMetadataPublicationToLocalRepository SKIPPED
:koap:publishMetadataPublicationToMavenRepository SKIPPED
:koap:publish SKIPPED
```
</details>